### PR TITLE
[Refactor] Separate PrimaryKeyTablePruneRule from RboTablePruneRule

### PR DESCRIPTION
**Pull request by satanson**
Refactor https://github.com/StarRocks/starrocks/pull/24843
RboTablePruneRule is poor readable, so separate PimaryKeyTablePruneRule from RboTablePruneRule, and some brain methods also simplified.

## What type of PR is this:
- [ ] BugFix
- [ ] Feature
- [ ] Enhancement
- [x] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Checklist:
- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.1
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 362 / 371 (97.57%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/pruner/PrimaryKeyUpdateTableRule.java|92|96|95.83%|[144, 147, 157, 169]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/join/JoinReorderCardinalityPreserving.java|38|39|97.44%|[283]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/pruner/CPJoinGardener.java|201|205|98.05%|[197, 345, 474, 489]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Utils.java|26|26|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/pruner/RboTablePruneRule.java|2|2|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 396 / 408 (97.06%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/pruner/CboTablePruneRule.java|34|37|91.89%|[157, 188, 194]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/pruner/PrimaryKeyUpdateTableRule.java|92|96|95.83%|[144, 147, 157, 169]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/join/JoinReorderCardinalityPreserving.java|38|39|97.44%|[283]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/pruner/CPJoinGardener.java|201|205|98.05%|[197, 345, 474, 489]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Utils.java|26|26|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/pruner/RboTablePruneRule.java|2|2|100.00%|[]|

**Review by kangkaisen — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 402 / 411 (97.81%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/pruner/PrimaryKeyUpdateTableRule.java|92|96|95.83%|[144, 147, 157, 169]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/pruner/CboTablePruneRule.java|42|44|95.45%|[148, 195]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/join/JoinReorderCardinalityPreserving.java|38|39|97.44%|[283]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/pruner/CPJoinGardener.java|199|201|99.00%|[479, 494]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Utils.java|26|26|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/pruner/RboTablePruneRule.java|2|2|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 402 / 411 (97.81%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/pruner/PrimaryKeyUpdateTableRule.java|92|96|95.83%|[144, 147, 157, 169]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/pruner/CboTablePruneRule.java|42|44|95.45%|[148, 195]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/join/JoinReorderCardinalityPreserving.java|38|39|97.44%|[283]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/pruner/CPJoinGardener.java|199|201|99.00%|[479, 494]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Utils.java|26|26|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/pruner/RboTablePruneRule.java|2|2|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Utils.java [hunk 1]
```diff
@@ -579,4 +589,42 @@ public static boolean canEliminateNull(Set<ColumnRefOperator> nullOutputColumnOp
         }
         return false;
     }
+
+    public static Stream<Integer> getIntStream(RoaringBitmap bitmap) {
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Utils.java:593]: get a better name? keyStream?

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 402 / 411 (97.81%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/pruner/PrimaryKeyUpdateTableRule.java|92|96|95.83%|[144, 147, 157, 169]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/pruner/CboTablePruneRule.java|42|44|95.45%|[148, 195]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/join/JoinReorderCardinalityPreserving.java|38|39|97.44%|[283]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/pruner/CPJoinGardener.java|199|201|99.00%|[479, 494]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Utils.java|26|26|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/pruner/RboTablePruneRule.java|2|2|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Utils.java (hunk 1 above)
Comment (satanson) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Utils.java:593]: bitmap can be considers as Set<unsigned Integer>, so getIntStream() return Stream<Integer>

**Review by Seaven — APPROVED** [review]

**Review by liuyehcf — APPROVED** [review]

**Review by ABingHuang — APPROVED** [review]

**Comment by satanson** [comment]
@mergifyio backport branch-3.1

**Pull request merged**